### ISSUE TITLE
Tidy treats boolean attributes inconsistently

### ISSUE THREAD

**Issue by jyasskin** [htacg/tidy-html5#142]
Given `test.html`:
```html
<!DOCTYPE html>
<html>
  <head>
    <title>Title</title>
    <script src="https://example.com/" async></script>
  </head>

  <body>
    <label><input type=checkbox checked name=cheese disabled> Cheese</label>
  </body>
</html>
```
`tidy --tidy-mark no --indent auto test.html` produces:
```html
<!DOCTYPE html>

<html>
<head>
  <title>Title</title>
  <script src="https://example.com/" async="">
</script>
</head>

<body>
  <label><input type="checkbox" checked name="cheese" disabled>
  Cheese</label>
</body>
</html>
```

The `<input>` element is formatted as I'd expect from the [boolean attributes](https://html.spec.whatwg.org/multipage/infrastructure.html#boolean-attributes) spec, but the [async](https://html.spec.whatwg.org/multipage/scripting.html#attr-script-async) attribute gets changed to include `=""`, which is unexpected.

**Comment by geoffmcl** [comment]
Hi Jeffrey (alias=jyasskin),

Yes, you are right...

This was a 'bug' introduced when adding
html5 support... it can be fixed by
changing one value in the source...

But I do not /YET/ have commit permission
on this repo, so can only fix it in my
own 'fork'... and in the past PRs have
been ignored... there are at least 14
outstanding... so have not done this...

So clone -
https://github.com/geoffmcl/tidy-fork

And then change to branch geoffmcl-restart...

$ git checkout geoffmcl-restart

And you should find it 'fixed'... and a LOT
of other features added, AND many other bug
fixes... I hope we soon get all this
together, for Tidy's sake... but it takes
time...

Be aware, my 'fork' only uses the cmake
generator build system... all others have
been removed...

Regard,
Geoff.

**Comment by balthisar** [comment]
We're working on commit strategy. Both Iionel (@skynet) and I have been
discussing some strategies, granting limited commit access, and so on.

The backlog of PR's will be addressed, but we have to be careful about
suddenly, blindly accepting the backlog without proper testing. I'm hopeful
that by the end of the week we will have a published strategy in place.

I recently accepted a token PR that couldn't have broken anything, at least
symbolically ending the PR drought.

The fork Geoff mentions *does* contain a significant portion of the PR
backlog. Any type of testing that you do in the meantime would make it that
much easier to commit back to the main repo.


On Tue, Jan 20, 2015 at 9:06 AM, Geoff McLane <notifications@github.com>
wrote:

> Hi Jeffrey (alias=jyasskin),
>
> Yes, you are right...
>
> This was a 'bug' introduced when adding
> html5 support... it can be fixed by
> changing one value in the source...
>
> But I do not /YET/ have commit permission
> on this repo, so can only fix it in my
> own 'fork'... and in the past PRs have
> been ignored... there are at least 14
> outstanding... so have not done this...
>
> So clone -
> https://github.com/geoffmcl/tidy-fork
>
> And then change to branch geoffmcl-restart...
>
> $ git checkout geoffmcl-restart
>
> And you should find it 'fixed'... and a LOT
> of other features added, AND many other bug
> fixes... I hope we soon get all this
> together, for Tidy's sake... but it takes
> time...
>
> Be aware, my 'fork' only uses the cmake
> generator build system... all others have
> been removed...
>
> Regard,
> Geoff.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/htacg/tidy-html5/issues/142#issuecomment-70588226>.
>



-- 
---
Jim Derry
Clinton Township, MI, USA
Nanjing, Jiangsu, China PRC

**Comment by skynet** [comment]
@geoffmcl - can you create a PR that will include only the small fix for this particular issue? Thanks.

**Comment by geoffmcl** [comment]
Hi Ionel,

Sorry do not know how to create a PR just for
that particular single issue...

But it is just changing one word, CH_PCDATA to
CH_BOOL, and optionally adding a comment, as
in the following diff extract -

diff -urw tidy-fork\src\attrs.c F:\Projects\tidy-fork\src\attrs.c
--- tidy-fork\src\attrs.c    Fri Aug 01 12:54:12 2014
+++ F:\Projects\tidy-fork\src\attrs.c    Mon Jan 19 13:36:31 2015
@@ -255,7 +260,7 @@
    { TidyAttr_URN,               "urn", CH_PCDATA    }, /* for <a>, 
never implemented */

    /* HTML5 */
-  { TidyAttr_ASYNC,             "async", CH_PCDATA    },
+  { TidyAttr_ASYNC,             "async", CH_BOOL      }, /* <script 
src="..." async> */
    { TidyAttr_AUTOCOMPLETE,      "autocomplete", CH_PCDATA    },
    { TidyAttr_AUTOFOCUS,         "autofocus", CH_PCDATA    },
    { TidyAttr_AUTOPLAY,          "autoplay", CH_PCDATA    },

Of course the line numbers may not be the same in
htacg/tidy-html5, but the full context is there...

Regards,
Geoff.

**Comment by geoffmcl** [comment]
Hi Jim, Iionel,

Yes, you certainly have to be carefully about
'blinding' accepting the backlog of PRs ;=()

In fact it seems PRs should have a 'sell-by-date',
since they are usually of the moment, and if the
base has moved on, then it is possible they are
either wrong or no longer needed... and can do
damage...

Meantime I have had a chance to commence regression
testing - that is run all the tests with pure CVS
(2009) tidy, and my tidy5 - and on the very first
test, 426885, found a REGRESSION had crept into
the base CVS code ;=(( thus is in all tidies since...

Tidy will in SOME cases no longer convert <center>
to a <div class..>, and in fact just DISCARDS it!!!
This happens if you also configure drop-font-tags...

The code is REAL simple and VERY clear -

static Bool Center2Div(...) {
  if (nodeIsCENTER(node)) {  // this is good, but
    if ( cfgBool(doc, TidyDropFontTags) ) { // why here?
      ... DiscardContainer(...); // YEEK, why this???
      // with no check that there is a <font...>
      return yes; // and now leaves...
    }
    // this does the <center> to a <div class..>
    // but ONLY if no drop-font-tags...
    RenameElem( doc, node, TidyTag_DIV );
    ... AddStyleProperty(...);
    return no;
}

What has font tags got to do with center???????
Seems the cfg check should just be removed...
So presently working hard on removing this BIG
BUG!!!

And not sure how you grant 'limited' commit rights?
I thought either a person has push permission or not.
And I would certainly be pleased to be one of the
maintainers...

And is there a dev-list we can use for discussion...
this seems strange having it inside an issue on a
particular repo ;=()

Anyway, ever onwards...

Regards,
Geoff.

**Comment by skynet** [comment]
HTML Tidy Advocacy Community Group Mailing List:

(public mailing list)[https://lists.w3.org/Archives/Public/public-htacg/]
(public mailing list for contributions)[https://lists.w3.org/Archives/Public/public-htacg-contrib/]
(internal mailing list for the HTML Tidy Advocacy Community Group.)[https://lists.w3.org/Archives/Member/internal-htacg/]

For this particular project I would definitely discuss everything right here though.

**Comment by geoffmcl** [comment]
Hi Ionel,

re: lists
1: https://lists.w3.org/Archives/Public/public-htacg/
2: https://lists.w3.org/Archives/Public/public-htacg-contrib/
3: https://lists.w3.org/Archives/Member/internal-htacg/

Well, especially since those lists seems totally
BLANK ;=))

So ok, here, and NOT there, or anywhere else!,
it seems, for now...

Fixed src/clean.c, of what seems like a regression,
just by removing the if (cfg == TidyDropFontTags)
code block from the if ( nodeIsCENTER(node) ) block...

All now pushed to my tidy-fork...

And progressing on the regression testing...

As always, the aim is the tidied version MUST look
the SAME in a browser... Quite slow since each
'test' requires a 'visual' browser inspection...

And really feel I must apologize to Jeffrey (jyasskin)
for hijacking this thread for a discussion ;=))))))

Regards,
Geoff.

**Comment by skynet** [comment]
Works for me. I hear about you updating your repo but my goal is to update this one. Have no idea what's the delta between this and yours.

**Comment by jyasskin** [comment]
Thanks for all the interest. :)

@geoffmcl Re "Sorry do not know how to create a PR just for that particular single issue", see https://help.github.com/articles/creating-a-pull-request/. You have to create a separate branch containing only the one change that fixes this issue.

**Comment by geoffmcl** [comment]
Hi Ionel,

 > deltas htacg vs tidy-fork

In a word, I would probably say 'massive' at this
point... and always growing...

I could generate a diff if this would help...

Or do a clone, and switch to the geoffmcl-restart
branch -
$ git checkout geoffmcl-restart

I will get around to merging this into master
soon...

But...

htacg has no ARIA support, quite big in itself...

htacg is missing a lot of fixes...

And for me the MOST IMPORTANT, htacg does NOT
use CMake...

I think it is important to settle on this last
soonest.

For me cmake is the ONLY way to go... it generates
just about ANY native build file set...

And to avoid all confusion I have blown away
Makefile, build/gmake. build/gnuauto,
build/msvc, build/msvc2010, and rpm, and adjusted
the README.md and READM.html accordingly...

And for unix this is as simple as -
$ cd make/cmake
$ cmake ../.. [options]
$ make
$ [sudo] make install (if desired)

An important option perhaps for unix is -
  -DBUILD_SHARED_LIB:BOOL=ON
It is off by default...

And it is just as simple for windows
$ cd make/cmake
$ cmake ..\..
$ cmake --build . --config Release
$ cmake --build . --config Release --target INSTALL (if desired)

*** BE AWARE ***, for my being able to fully test
my fork while keeping standard tidy available
I have **MESSED** with the library and binary names...
using lib tidy-static or -shared, and tidy5...

The idea being only when fully satisfied should
the names be returned to the simple... libtidy,
and tidy... and thus overwrite any existing install...

This is a simple adjustment in the single CMakeLists.txt

So, how do we go about bringing htacg into this
century? ;=))))

With htacg commit rights I would be prepared to
do the initial merge, probably by this weekend...

And finally, there is the question of a version,
again a CMakeTests.txt adjustment only...

Regards,
Geoff.

**Comment by skynet** [comment]
Thanks, Geoff. can you commit into a dev branch?

**Comment by geoffmcl** [comment]
No problem...

**Comment by geoffmcl** [comment]
Hi Jeffrey,

Thanks for the info on generating a PR
for a single issue...

But I thought that would be the case - to create a
separate branch - but I was not prepared to do
that just for a one word change ;=))

Will keep it in mind for perhaps 'bigger'
things...

But my biggest hope is that we can get htacg
up-to-date and I can forget my tidy-fork...

It was after all only generated because there was
absolutely NO activity on tidy's health... for
years and Years and YEARS!!!

Regards,
Geoff.

**Comment by balthisar** [comment]
Geoff, I'll try to pull yours in the next couple of days. It's the most
logical approach. We're still labelled "experimental" and so Ionel has
convinced me that the risk is low.

I'm a little worried about the other build systems that were removed. This
could be good, much needed pruning (especially if they didn't work), or it
could break someone's build process (probably unlikely given how static
Tidy has been).

Not sure if you had a chance to clean up my customized version strings. I
don't mean to be assigning tasks, and I can clean them up before merging.

On the issue of merging, what I'll probably do is just replace the entire
existing repo. I think this gives us the best chance of avoiding conflicts.
The new commit will properly show all of the diffs from file to file,
HOWEVER --- we're going to lose some of the PR history. This has the
potential to hurt a lot of contributors' feelings. Their names and PR
numbers will still show up in the commit history (I added these to the
commits), but they won't show up in git's magic who-contributed-what
history.

It sounds like you've mastered how to merge all of the changes without the
conflicts I kept getting. If I give you commit access, would you be willing
to merge in the current -restart branch into master? Maybe give it a trial
run in /htacg in a test branch that can be deleted later.

If not, I'll try to do it before the weekend.

As a contributor with a demonstrated commitment to Tidy I'll be happy to
let you keep the commit access, but this initial merge and commit should be
a one-time deal only, until we've decided on proper change control process.




On Wed, Jan 21, 2015 at 2:57 AM, Geoff McLane <notifications@github.com>
wrote:

> Hi Jeffrey,
>
> Thanks for the info on generating a PR
> for a single issue...
>
> But I thought that would be the case - to create a
> separate branch - but I was not prepared to do
> that just for a one word change ;=))
>
> Will keep it in mind for perhaps 'bigger'
> things...
>
> But my biggest hope is that we can get htacg
> up-to-date and I can forget my tidy-fork...
>
> It was after all only generated because there was
> absolutely NO activity on tidy's health... for
> years and Years and YEARS!!!
>
> Regards,
> Geoff.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/htacg/tidy-html5/issues/142#issuecomment-70710611>.
>



-- 
---
Jim Derry
Clinton Township, MI, USA
Nanjing, Jiangsu, China PRC

**Comment by balthisar** [comment]
@geoffmcl I've given commit access and created a repo at /htacg/geoff-test. If you would like to do some merge testing there, we can talk about bringing /htacg/tidy-html5 up to date.

Thanks for your support.

**Comment by geoffmcl** [comment]
Hi Jim,

===
* merging

The BIG trick with git merging is to be prepared
to do it SEVERAL times ;=))

And I always keep handy a clean clone, like say
tidy-html5-org, so I can refer back to it BEFORE
git tried...

On the initial merge it is ALWAYS likely there will
be a number of conflicts. Now conflicts can be of
several types -

(a) A 'real' merge conflict. This is where both
versions have been modified around the same area
and git wants you to manually decided WHICH code
is to take precedence, or how to combine the
efforts/lines...

This has to be done MANUALLY, but since git
places big markers around the two blocks it is
usually not so difficult. Perhaps tedious...

Additionally I have a wonderful tool called
WinMerge which loads the two files and gives a
two pane view, color coding the lines that
are different... Making easy to see what has
been changed, where...

But in reality between my fork and html-tidy5
there should be none, or few of these since
little has been done to html-tidy5 since I
first cloned from it...

(b) A 'false' merge conflict. This is where
only ONE files has really been modified, but
during the modification some white spaces were
added or removed on other lines... ie lines other
than the NEW lines added...

Git gets into a real mess with this, and will
report both files changed, when in fact my
WinMerge tool, which IGNORES white space changes
in lines, shows me that only a bunch of lines
have been added.

Really no merge problem at all...

In most cases like this it is a simple case of
COPYING the new file over the old file...

(c) Really false, false merge conflicts. And
I think this is again related to white space
changes...

That is my WinMerge tool, which as stated IGNORES
white space changes, reports the two files as
the SAME! As does diff -w...

Again just copy the new over the old... And things
go better if this is done before starting the
merge again on a new clean clone...

(d) Where files have been deleted in one,
but not deleted in the other.

I suppose in the interests of safety git
will leave the file there, and report a
conflict.

In this case the trick, like (c) above, is to
trash the current merge repo, and clone again,
and do the same deletes BEFORE attempting the
next merge, and git will be happy...

(e) Maybe others I have not experienced ;=))

As stated, always a bit TEDIOUS, but doable...

As you point out the only real reason to be
doing a merge is to PRESERVE the files git
history - who changed what, when, where...

Of course if you do NOT care about history,
then just a 100% replacement is possible.

And as is my habit, I add notes to myself, and
others, on my site. This is a recent tidy merge -

  http://geoffair.org/unix/git.htm#merge1

===
* cmake

I guess you are not too familiar with cmake?

You do understand cmake is not a build system
in itself, but a configuration and generator
of 'native' build systems...

Running cmake without a command will show
the list of generators available on your
platform, which will include Unix Makefiles,
those processed by the venerable 'make'...

And as a tidy build has always been cmake
supports out-of-source builds, making
cleanups, restarts, ... easy...

And is this not a new generation of tidy,
so why not simplify life for everyone,
and settle on ONLY cmake?

And concerning version.h I have moved the
'versioning' into CMakeLists.txt... which
generates a RELEASE_DATE string macro...

So this can be the current date, but I
was also considering the so library version,
since cmake makes building shared library
just as easy... which unix loves, but not
so much for windows, where they create
extra hassle...

Please, let's go 100% cmake...

===
* push

I really do not understand what you are
doing concerning permissions???

I do not like blowing my own trumpet,
but I think I submitted my first patch
to tidy in 2006 or earlier... and continued
ever since then... getting close to 10
years... check out the tidy-develop mailing
list... The CVS code, and thus htacq/html-tidy5
has MANY of MY patches...

I never push for CVS access there since arnaud
was always there, although Charles gave
me access to the TidyLib Binding...

What does giving me access to an empty
htacq/geoff-test repo do?

I had already agreed that the merge
would be done in a branch... so master
would be UNTOUCHED... and if you did not
like the resulting branch, it could be
trashed.

Anyway, if I can not be trusted with
git push access to htacq/html-tidy5, that is
fine, and I guess it is up to you to do
the merge, if you want...

I will return to quietly working on my
own fork...

Regards,
Geoff.

**Comment by skynet** [comment]
Maybe we can setup some coding style to be applied to all commits, prior to PRs, especially in regard to white space. We shouldn't have such problems, but it's normal, considering that everyone is using their editor of choice.

**Comment by skynet** [comment]
IMHO we can't commit large changes to master and then consider this ready to be used in production. Let's just just setup a dev branch and advise people to deploy only for development or testing.

**Comment by petdance** [comment]
Agreed: There needs to be a dev branch.  People coming in to look at the project will expect that master is deployable.

**Comment by geoffmcl** [comment]
Hi,

Just for fun did a merge from geoffmcl/tidy-fork,
geoffmcl-restart branch to htacg/tidy-html5,
geoffmcl-restart branch, and surprise, SURPRISE!
none of the predicted merge problems were there ;=))

Got exactly one(1) conflict, and that was in the
README.md which was easily fixed...

And that geoffmcl-retart branch can be easily
renamed 'dev' or any other name...
$ git branch -m geoffmcl-restart <newname>

And I need to fix up some of the info messages
in tidy.c, since it refers to my repo, etc... but
just simple message changes back to near the
original...

Then if I had push rights... it would be done,
and ready for people to try ;=))

I have continued the full regression testing,
and so far except for the first test, which
I have now fixed, at least the first 20 or so
are fine... ongoing...

Regards,
Geoff.

**Comment by geoffmcl** [comment]
Hi,

Have now returned the tidy.c info messages to their
former glory, but it seems some URL changes
are needed...

Yes, it seems they are re-directed, but also
seems no reason not to put the correct current URL.

And should we leave validator as 'nu', now that
the main validator handles both, or we could put
both...

==================================
 > tidy5 -h

tidy5 [options...] [file...] [options...] [file...]
Utility to clean up and pretty print HTML/XHTML/XML

This is an HTML5-aware experimental fork of HTML Tidy.
2015/01/20

<skipped unchanged options text>

For more information on this HTML5-aware experimental fork of Tidy,
see http://w3c.github.com/tidy-html5/

For more information on HTML, see the following:

   HTML: Edition for Web Authors (the latest HTML specification)
   http://dev.w3.org/html5/spec-author-view

   HTML: The Markup Language (an HTML language reference)
   http://dev.w3.org/html5/markup/

File bug reports at https://github.com/w3c/tidy-html5/issues/
or send questions and comments to html-tidy@w3.org

Validate your HTML documents using the W3C Nu Markup Validator:

   http://validator.w3.org/nu/

==================================

For the present have set the version output as -

 > tidy5 --version
HTML Tidy for HTML5 (experimental) for Windows 2015/01/20

Of course that says 'Unix' if compiled in unix/linux...

Seek advice on what to change... if anything...

Regards,
Geoff.

**Comment by balthisar** [comment]
@geoffmcl 

>* push
>
>I really do not understand what you are
>doing concerning permissions???
> ...
>What does giving me access to an empty
>htacq/geoff-test repo do?

>I had already agreed that the merge
>would be done in a branch... so master
>would be UNTOUCHED... and if you did not
>ike the resulting branch, it could be
>trashed.

I missed that part of the discussion while I was sleeping in my timezone. I thought that a trash repo would be a convenience to you.

However I did give you commit access, and although I didn't say outright that it includes access to the tidy-html5 repo, it includes access to the tidy-html5 repo.

**Comment by balthisar** [comment]
@geoffmcl said:
>And is this not a new generation of tidy,
>so why not simplify life for everyone,
>and settle on ONLY cmake?
> ...
> Please, let's go 100% cmake...

I'm willing to support that, but I'm only one person; can we try an experiment so: post this suggestion to mailto:public-htacg@w3.org (https://lists.w3.org/Archives/Public/public-htacg/)?

I know the list hasn't seen any activity yet, but if we're going to be a serious organization I suggest that this type of debate be open to everyone at HTACG who doesn't necessarily check out these issues threads. Since literally anyone in the public can join HTACG, it's not exclusionary and offers full transparency.

I could post to the list, but I don't want to steal your thunder...

**Comment by geoffmcl** [comment]
Hi Jim,

Two problems -

1: Can not push the htacg/tidy-html5. Get error
Permission to htacg/tidy-html5.git denied to geoffmcl
so although ready unable to push develop-500 branch.

2: Can NOT place a message on either public-htacg,
or public-htacg-contrib lists. Have checked
allowing post to be public, and on a second try
it advised already done (approved)

So have returned here for help.

Regards,
Geoff.

**Comment by balthisar** [comment]
Confirmed fixed in develop-500. I will close this. Thanks, Geoff.
